File: scripts/parse-events.py
#!/usr/bin/env python3
"""Парсер событий iron-star.com/event/ → .md файлы в src/content/events/"""

import re, os, sys
from datetime import datetime

import requests
from bs4 import BeautifulSoup

def parse_event_text(text: str) -> dict | None:
    """Парсит строку карточки события в структуру (пошаговый разбор)."""
    text = text.strip()

    # 1. Находим дату — разделитель между названием и остальным
    date_pattern = re.compile(r'\d{2}\.\d{2}\.\d{4}')
    date_m = date_pattern.search(text)
    if not date_m:
        print(f'  SKIP (no date): {text[:80]}', file=sys.stderr)
        return None

    name = text[:date_m.start()].strip()
    date = datetime.strptime(date_m.group(), '%d.%m.%Y')
    after_date = text[date_m.end():].strip()

    # 2. Ищем статус
    status_map = {
        'осталось мало слотов': 'low',
        'продано': 'sold_out',
        'лист ожидания': 'waitlist',
    }
    status_keywords = list(status_map.keys())
    status_pattern = re.compile('|'.join(re.escape(k) for k in status_keywords), re.IGNORECASE)

    status = 'open'
    status_m = status_pattern.search(after_date)

    if status_m:
        # Всё до статуса: "ГОРОД НАЗВАНИЕ"
        before_status = after_date[:status_m.start()].strip()
        # Всё после статуса: дистанции
        distances_str = after_date[status_m.end():].strip()
        status = status_map[status_m.group().lower()]
    else:
        # Нет статуса — after_date = "ГОРОД НАЗВАНИЕ ДИСТАНЦИИ"
        before_status = after_date
        distances_str = ''

    # 3. Отделяем город: «before_status» кончается на название события
    # Находим последнее вхождение названия
    name_pos = before_status.rfind(name)
    if name_pos >= 0:
        city = before_status[:name_pos].strip()
        # Убираем название + возможные повторы дистанций (если нет статуса)
        rest = before_status[name_pos + len(name):].strip()
        if rest and not distances_str:
            # rest = возможные дистанции
            distances_str = rest
    else:
        # Название не нашлось (редкий случай, например разный регистр)
        # Берём первое слово после даты как город
        parts = before_status.split()
        city = parts[0] if parts else before_status

    # 4. Парсим дистанции
    dist_re = re.compile(r'([\d,]+)\s*км')
    if not distances_str:
        # Ищем дистанции во всей строке после даты (fallback)
        distances_str = after_date
    distances = [float(d.replace(',', '.')) for d in dist_re.findall(distances_str)]

    if not distances:
        print(f'  SKIP (no distances): {text[:80]}', file=sys.stderr)
        return None

    # 5. Определяем sportType по количеству и порядку дистанций + названию
    name_upper = name.upper()
    if len(distances) == 1:
        if any(kw in name_upper for kw in ['SWIM', 'ПЛАВ']):
            sport_type = 'swim'
            mapped = {'swim': distances[0]}
        else:
            sport_type = 'run'
            mapped = {'run': distances[0]}
    elif len(distances) == 2:
        if 'SWIMRUN' in name_upper or 'СВИМРАН' in name_upper:
            sport_type = 'swimrun'
            mapped = {'swim': distances[0], 'run': distances[1]}
        elif 'AQUATHLON' in name_upper or 'АКВАТЛОН' in name_upper:
            sport_type = 'aquathlon'
            mapped = {'swim': distances[0], 'run': distances[1]}
        else:
            sport_type = 'aquathlon'
            mapped = {'swim': distances[0], 'run': distances[1]}
    elif len(distances) == 3:
        sport_type = 'triathlon'
        mapped = {'swim': distances[0], 'bike': distances[1], 'run': distances[2]}
    else:
        print(f'  SKIP: unexpected distance count {len(distances)}: {text[:80]}', file=sys.stderr)
        return None

    return {
        'title': name,
        'date': date,
        'city': city,
        'sportType': sport_type,
        'status': status,
        'distances': mapped,
    }


def fetch_events() -> list[str]:
    """Получает список текстов карточек событий с iron-star.com/event/"""
    url = 'https://iron-star.com/event/'
    headers = {
        'User-Agent': 'Mozilla/5.0 (compatible; PlanExBot/1.0; +https://dimino.me)',
    }
    resp = requests.get(url, headers=headers, timeout=30)
    resp.raise_for_status()

    soup = BeautifulSoup(resp.text, 'html.parser')

    # Ищем ссылки, содержащие дату в тексте
    date_pattern = re.compile(r'\d{2}\.\d{2}\.\d{4}')

    events = []
    seen = set()
    for a in soup.find_all('a', href=True):
        text = a.get_text(separator=' ', strip=True)
        if not text:
            continue
        # Пропускаем кнопки и системные ссылки
        if text.lower() in ('регистрация', 'подробнее', 'лист ожидания', 'eng', 'rus',
                            'фильтр соревнований', 'личный кабинет', 'архив результатов',
                            'помощь', 'подписаться на рассылку'):
            continue
        # Пропускаем дубликаты
        if text in seen:
            continue
        # Проверяем, что текст содержит дату и "км"
        if date_pattern.search(text) and 'км' in text.lower():
            seen.add(text)
            events.append(text)

    return events


def slugify(title: str, date: datetime) -> str:
    """Генерирует slug для имени файла."""
    name_slug = re.sub(r'[^a-zа-яё0-9]+', '-', title.lower()).strip('-')
    date_slug = date.strftime('%Y-%m-%d')
    return f'{date_slug}-{name_slug}'


def generate_md(event: dict) -> str:
    """Генерирует содержимое .md файла."""
    distances = event['distances']
    dist_lines = '\n'.join(f'    {k}: {v}' for k, v in distances.items())

    # Формируем тело без заголовка h1 (данные уже в frontmatter)
    body = ''
    for k, v in distances.items():
        label = {'swim': '🏊 Плавание', 'bike': '🚴 Вело', 'run': '🏃 Бег'}.get(k, k)
        body += f'- {label}: {v} км\n'

    return f'''---
title: "{event['title']}"
date: {event['date'].strftime('%Y-%m-%d')}
city: "{event['city']}"
sportType: {event['sportType']}
status: {event['status']}
distances:
{dist_lines}
---

**Дата:** {event['date'].strftime('%d.%m.%Y')}  
**Город:** {event['city']}  
**Тип:** {event['sportType']}  
**Статус слотов:** {event['status']}

## Дистанции

{body}
[Регистрация на iron-star.com](https://iron-star.com/event/)
'''


def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_dir = os.path.dirname(script_dir)
    events_dir = os.path.join(project_dir, 'src', 'content', 'events')

    os.makedirs(events_dir, exist_ok=True)

    print('Fetching events from iron-star.com/event/ ...')
    texts = fetch_events()
    print(f'Found {len(texts)} event cards')

    parsed = 0
    saved = 0

    for text in texts:
        event = parse_event_text(text)
        if not event:
            continue
        parsed += 1

        slug = slugify(event['title'], event['date'])
        md_content = generate_md(event)
        filepath = os.path.join(events_dir, f'{slug}.md')

        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(md_content)
        print(f'  ✓ {slug}.md')
        saved += 1

    print(f'\nDone: {parsed} parsed, {saved} saved to {events_dir}')


if __name__ == '__main__':
    main()

File: scripts/parse-events.mjs
#!/usr/bin/env node
/**
 * Парсер событий iron-star.com/event/ → .md файлы в src/content/events/
 * Использует Playwright, так как сайт — Nuxt SPA.
 *
 * Собирает: title, date, city, sportType, status, distances, sourceUrl, prices
 *
 * Usage: node scripts/parse-events.mjs [--skip-prices]
 */

import { chromium } from 'playwright';
import { readFile, writeFile, mkdir, readdir, unlink } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_DIR = join(__dirname, '..');
const EVENTS_DIR = join(PROJECT_DIR, 'src', 'content', 'events');

const SKIP_PRICES = process.argv.includes('--skip-prices');

/**
 * Определяет sportType по названию и дистанциям.
 */
function detectSportType(title, dists) {
  const nameUp = title.toUpperCase();

  if (dists.length === 1) {
    if (/SWIM|ПЛАВ/.test(nameUp)) return { type: 'swim', dists: { swim: dists[0] } };
    if (/RUN|БЕГ|IRONLADY|MANSTAR|STARKIDS|5K/.test(nameUp))
      return { type: 'run', dists: { run: dists[0] } };
    return { type: 'run', dists: { run: dists[0] } };
  }

  if (dists.length === 2) {
    if (/SWIMRUN|СВИМРАН/.test(nameUp))
      return { type: 'swimrun', dists: { swim: dists[0], run: dists[1] } };
    if (/AQUATHLON|АКВАТЛОН/.test(nameUp))
      return { type: 'aquathlon', dists: { swim: dists[0], run: dists[1] } };
    // Детские старты, relay — swim/run
    if (/RELAY/.test(nameUp)) return { type: 'swim', dists: { swim: dists[0] } };
    return { type: 'aquathlon', dists: { swim: dists[0], run: dists[1] } };
  }

  if (dists.length === 3) {
    return { type: 'triathlon', dists: { swim: dists[0], bike: dists[1], run: dists[2] } };
  }

  return null;
}

/**
 * Определяет статус слотов.
 */
function detectStatus(statusText) {
  const s = (statusText || '').toLowerCase();
  if (s.includes('мало')) return 'low';
  if (s.includes('продано')) return 'sold_out';
  if (s.includes('ожидани')) return 'waitlist';
  return 'open';
}

/**
 * Парсит дистанцию из строки вида "1,93 км" или "1.93 км"
 */
function parseDistance(text) {
  const cleaned = text.replace(/\s*км\s*/i, '').replace(',', '.');
  const val = parseFloat(cleaned);
  return isNaN(val) ? 0 : val;
}

/**
 * Slug для имени файла
 */
function slugify(title, dateStr) {
  const nameSlug = title
    .toLowerCase()
    .replace(/[^a-zа-яё0-9]+/g, '-')
    .replace(/^-|-$/g, '');
  return `${dateStr}-${nameSlug}`;
}

/**
 * Парсит цены с отдельной страницы события.
 * Возвращает массив { category, levels: [{ level, price }] }
 */
async function parsePrices(page, eventUrl) {
  if (SKIP_PRICES) return [];

  try {
    await page.goto(eventUrl, { waitUntil: 'networkidle', timeout: 15000 });

    // Ищем блок "СТОИМОСТЬ УЧАСТИЯ"
    const prices = await page.evaluate(() => {
      const result = [];
      const allText = document.body.innerText;

      // Если нет раздела стоимости — возвращаем пустой
      if (!allText.includes('СТОИМОСТЬ УЧАСТИЯ')) return result;

      // Ищем все段落 с ценами
      // Формат: "Индивидуальное участие" → уровни, "Эстафета" → уровни
      const priceRegex = /(\d+)\s*уровень:\s*([\d\s]+)\s*₽/gi;
      const lines = allText.split('\n');

      let currentCategory = 'Индивидуальное участие';
      const categories = {};

      for (const line of lines) {
        const catMatch = line.match(/^(Индивидуальное участие|Эстафета|Командное участие)/i);
        if (catMatch) {
          currentCategory = catMatch[1];
          continue;
        }

        const match = /(\d+)\s*уровень:\s*([\d\s]+)\s*₽/i.exec(line);
        if (match) {
          if (!categories[currentCategory]) categories[currentCategory] = [];
          categories[currentCategory].push({
            level: parseInt(match[1]),
            price: parseInt(match[2].replace(/\s/g, '')),
          });
        }
      }

      for (const [cat, levels] of Object.entries(categories)) {
        result.push({ category: cat, levels });
      }

      return result;
    });

    return prices;
  } catch (err) {
    console.log(`    ⚠ Price parse failed: ${err.message}`);
    return [];
  }
}

/**
 * Генерирует .md контент
 */
function generateMd(event) {
  const { title, dateStr, city, sportType, status, distances, sourceUrl, prices } = event;
  const distKeys = Object.entries(distances);

  const frontmatterDists = distKeys.map(([k, v]) => `    ${k}: ${v}`).join('\n');

  const bodyLines = distKeys
    .map(([k, v]) => {
      const labels = { swim: '🏊 Плавание', bike: '🚴 Вело', run: '🏃 Бег' };
      return `- ${labels[k] || k}: ${v} км`;
    })
    .join('\n');

  // Prices in YAML
  let pricesYaml = '';
  if (prices && prices.length > 0) {
    pricesYaml =
      '\nprices:\n' +
      prices
        .map(
          (p) =>
            `  - category: "${p.category}"\n    levels:\n${p.levels.map((l) => `      - { level: ${l.level}, price: ${l.price} }`).join('\n')}`,
        )
        .join('\n');
  }

  // Prices in body
  let pricesBody = '';
  if (prices && prices.length > 0) {
    pricesBody = '\n\n## Стоимость участия\n';
    for (const cat of prices) {
      if (prices.length > 1) pricesBody += `\n**${cat.category}:**\n`;
      for (const l of cat.levels) {
        pricesBody += `- ${l.level} уровень: ${l.price.toLocaleString('ru-RU')} ₽\n`;
      }
    }
  }

  const sourceUrlStr = sourceUrl ? `\nsourceUrl: "${sourceUrl}"` : '';

  return `---
title: "${title}"
date: ${dateStr}
city: "${city}"
sportType: ${sportType}
status: ${status}
distances:
${frontmatterDists}${sourceUrlStr}${pricesYaml}
---

**Дата:** ${dateStr.replace(/-/g, '.')}
**Город:** ${city}
**Тип:** ${sportType}
**Статус слотов:** ${status}

## Дистанции

${bodyLines}
${pricesBody}
${sourceUrl ? `[Регистрация на iron-star.com →](${sourceUrl})` : '[Регистрация на iron-star.com](https://iron-star.com/event/)'}
`;
}

async function main() {
  await mkdir(EVENTS_DIR, { recursive: true });

  console.log('Launching browser...');
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  try {
    console.log('Navigating to iron-star.com/event/ ...');
    await page.goto('https://iron-star.com/event/', {
      waitUntil: 'networkidle',
      timeout: 30000,
    });

    // Ждём загрузки карточек
    await page.waitForSelector('.event-item-wrap', { timeout: 10000 });

    // Скроллим до конца для загрузки всех событий (виртуальный скроллинг)
    let prevCount = 0;
    for (let i = 0; i < 20; i++) {
      await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
      await page.waitForTimeout(500);
      const count = await page.$$eval('.event-item-wrap', (els) => els.length);
      if (count === prevCount && count > 0) break;
      prevCount = count;
    }

    // Извлекаем данные с главной страницы
    const events = await page.$$eval('.event-item-wrap', (els) =>
      els.map((el) => {
        const link = el.querySelector('a.event-item');
        const title = el.querySelector('.title')?.textContent?.trim() || '';
        const date = el.querySelector('.date')?.textContent?.trim() || '';
        const place = el.querySelector('.place')?.textContent?.trim() || '';
        // Статус в div внутри .event-image
        const imageDiv = el.querySelector('.event-image');
        const statusDiv = imageDiv?.querySelector('div');
        const statusText = statusDiv?.textContent?.trim() || '';
        const distances = Array.from(el.querySelectorAll('.distance')).map((d) =>
          d.textContent.trim(),
        );
        return { title, date, place, statusText, distances, href: link?.href || '' };
      }),
    );

    console.log(`Found ${events.length} event cards\n`);

    let saved = 0;
    let skipped = 0;

    for (const raw of events) {
      const { title, date: dateRaw, place: city, statusText, distances: distStrs, href } = raw;

      // Парсим дату ДД.ММ.ГГГГ → YYYY-MM-DD
      const dateMatch = dateRaw.match(/^(\d{2})\.(\d{2})\.(\d{4})$/);
      if (!dateMatch) {
        console.log(`  SKIP (bad date): ${title} ${dateRaw}`);
        skipped++;
        continue;
      }
      const dateStr = `${dateMatch[3]}-${dateMatch[2]}-${dateMatch[1]}`;

      // Парсим дистанции
      const dists = distStrs.map(parseDistance).filter((d) => d > 0);
      if (dists.length === 0) {
        console.log(`  SKIP (no distances): ${title}`);
        skipped++;
        continue;
      }

      const sport = detectSportType(title, dists);
      if (!sport) {
        console.log(`  SKIP (unknown sport): ${title} ${dists.length} dists`);
        skipped++;
        continue;
      }

      const status = detectStatus(statusText);
      const sourceUrl = href || '';
      const slug = slugify(title, dateStr);

      console.log(`  Parsing prices for ${title}...`);
      const prices = await parsePrices(page, sourceUrl);

      const md = generateMd({
        title,
        dateStr,
        city,
        sportType: sport.type,
        status,
        distances: sport.dists,
        sourceUrl,
        prices: prices.length > 0 ? prices : undefined,
      });

      const filepath = join(EVENTS_DIR, `${slug}.md`);
      await writeFile(filepath, md, 'utf-8');
      console.log(`  ✓ ${slug}.md (${sport.type}, ${status}, ${prices.length} price cats)`);
      saved++;
    }

    console.log(`\nDone: ${saved} saved, ${skipped} skipped → ${EVENTS_DIR}`);
  } finally {
    await browser.close();
  }
}

main().catch((err) => {
  console.error('FATAL:', err);
  process.exit(1);
});
